File: ffi/diplomat/src/date.rs
// This file is part of ICU4X. For terms of use, please see the file
// called LICENSE at the top level of the ICU4X source tree
// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).

#[diplomat::bridge]
pub mod ffi {
    use alloc::boxed::Box;
    use alloc::sync::Arc;
    use core::fmt::Write;
    use diplomat_runtime::DiplomatResult;
    use icu_calendar::types::IsoWeekday;
    use icu_calendar::AnyCalendar;
    use icu_calendar::{Date, Iso};

    use crate::calendar::ffi::ICU4XCalendar;
    use crate::errors::ffi::ICU4XError;

    #[diplomat::enum_convert(IsoWeekday)]
    pub enum ICU4XIsoWeekday {
        Monday = 1,
        Tuesday,
        Wednesday,
        Thursday,
        Friday,
        Saturday,
        Sunday,
    }
    #[diplomat::opaque]
    #[diplomat::transparent_convert]
    /// An ICU4X Date object capable of containing a ISO-8601 date
    #[diplomat::rust_link(icu::calendar::Date, Struct)]
    pub struct ICU4XIsoDate(pub Date<Iso>);

    impl ICU4XIsoDate {
        /// Creates a new [`ICU4XIsoDate`] from the specified date and time.
        #[diplomat::rust_link(icu::calendar::Date::new_iso_date, FnInStruct)]
        pub fn try_new(
            year: i32,
            month: u8,
            day: u8,
        ) -> DiplomatResult<Box<ICU4XIsoDate>, ICU4XError> {
            Date::new_iso_date(year, month, day)
                .map(|dt| Box::new(ICU4XIsoDate(dt)))
                .map_err(Into::into)
                .into()
        }

        /// Convert this date to one in a different calendar
        #[diplomat::rust_link(icu::calendar::Date::to_calendar, FnInStruct)]
        pub fn to_calendar(&self, calendar: &ICU4XCalendar) -> Box<ICU4XDate> {
            Box::new(ICU4XDate(self.0.to_calendar(calendar.0.clone())))
        }

        #[diplomat::rust_link(icu::calendar::Date::to_any, FnInStruct)]
        pub fn to_any(&self) -> Box<ICU4XDate> {
            Box::new(ICU4XDate(self.0.to_any().wrap_calendar_in_arc()))
        }

        /// Returns the 1-indexed day in the month for this date
        #[diplomat::rust_link(icu::calendar::Date::day_of_month, FnInStruct)]
        pub fn day_of_month(&self) -> u32 {
            self.0.day_of_month().0
        }

        /// Returns the day in the week for this day
        #[diplomat::rust_link(icu::calendar::Date::day_of_week, FnInStruct)]
        pub fn day_of_week(&self) -> ICU4XIsoWeekday {
            self.0.day_of_week().into()
        }

        /// Returns 1-indexed number of the month of this date in its year
        #[diplomat::rust_link(icu::calendar::Date::month, FnInStruct)]
        pub fn month(&self) -> u32 {
            self.0.month().ordinal
        }

        /// Returns the year number for this date
        #[diplomat::rust_link(icu::calendar::Date::year, FnInStruct)]
        pub fn year(&self) -> i32 {
            self.0.year().number
        }

        /// Returns the number of months in the year represented by this date
        #[diplomat::rust_link(icu::calendar::Date::months_in_year, FnInStruct)]
        pub fn months_in_year(&self) -> u8 {
            self.0.months_in_year()
        }

        /// Returns the number of days in the month represented by this date
        #[diplomat::rust_link(icu::calendar::Date::days_in_month, FnInStruct)]
        pub fn days_in_month(&self) -> u8 {
            self.0.days_in_month()
        }

        /// Returns the number of days in the year represented by this date
        #[diplomat::rust_link(icu::calendar::Date::days_in_year, FnInStruct)]
        pub fn days_in_year(&self) -> u32 {
            self.0.days_in_year()
        }
    }

    #[diplomat::opaque]
    #[diplomat::transparent_convert]
    /// An ICU4X Date object capable of containing a date and time for any calendar.
    #[diplomat::rust_link(icu::calendar::Date, Struct)]
    pub struct ICU4XDate(pub Date<Arc<AnyCalendar>>);

    impl ICU4XDate {
        /// Creates a new [`ICU4XDate`] representing the ISO date and time
        /// given but in a given calendar
        #[diplomat::rust_link(icu::calendar::Date::new_from_iso, FnInStruct)]
        pub fn try_new_from_iso_in_calendar(
            year: i32,
            month: u8,
            day: u8,
            calendar: &ICU4XCalendar,
        ) -> DiplomatResult<Box<ICU4XDate>, ICU4XError> {
            let cal = calendar.0.clone();
            Date::new_iso_date(year, month, day)
                .map(|dt| Box::new(ICU4XDate(dt.to_calendar(cal))))
                .map_err(Into::into)
                .into()
        }

        /// Creates a new [`ICU4XDate`] from the given codes, which are interpreted in the given calendar system
        #[diplomat::rust_link(icu::calendar::Date::new_from_codes, FnInStruct)]
        pub fn try_new_from_codes_in_calendar(
            era_code: &str,
            year: i32,
            month_code: &str,
            day: u8,
            calendar: &ICU4XCalendar,
        ) -> DiplomatResult<Box<ICU4XDate>, ICU4XError> {
            let era = try_icu4x!(era_code.parse());
            let month = try_icu4x!(month_code.parse());
            let cal = calendar.0.clone();
            Date::new_from_codes(era, year, month, day, cal)
                .map(|dt| Box::new(ICU4XDate(dt)))
                .map_err(Into::into)
                .into()
        }

        /// Convert this date to one in a different calendar
        #[diplomat::rust_link(icu::calendar::Date::to_calendar, FnInStruct)]
        pub fn to_calendar(&self, calendar: &ICU4XCalendar) -> Box<ICU4XDate> {
            Box::new(ICU4XDate(self.0.to_calendar(calendar.0.clone())))
        }

        /// Converts this date to ISO
        #[diplomat::rust_link(icu::calendar::Date::to_iso, FnInStruct)]
        pub fn to_iso(&self) -> Box<ICU4XIsoDate> {
            Box::new(ICU4XIsoDate(self.0.to_iso()))
        }

        /// Returns the 1-indexed day in the month for this date
        #[diplomat::rust_link(icu::calendar::Date::day_of_month, FnInStruct)]
        pub fn day_of_month(&self) -> u32 {
            self.0.day_of_month().0
        }

        /// Returns the day in the week for this day
        #[diplomat::rust_link(icu::calendar::Date::day_of_week, FnInStruct)]
        pub fn day_of_week(&self) -> ICU4XIsoWeekday {
            self.0.day_of_week().into()
        }
        /// Returns 1-indexed number of the month of this date in its year
        ///
        /// Note that for lunar calendars this may not lead to the same month
        /// having the same ordinal month across years; use month_code if you care
        /// about month identity.
        #[diplomat::rust_link(icu::calendar::Date::month, FnInStruct)]
        pub fn ordinal_month(&self) -> u32 {
            self.0.month().ordinal
        }

        /// Returns the month code for this date. Typically something
        /// like "M01", "M02", but can be more complicated for lunar calendars.
        #[diplomat::rust_link(icu::calendar::Date::month, FnInStruct)]
        pub fn month_code(
            &self,
            write: &mut diplomat_runtime::DiplomatWriteable,
        ) -> DiplomatResult<(), ICU4XError> {
            let code = self.0.month().code;
            let result = write.write_str(&code.0).map_err(Into::into).into();
            write.flush();
            result
        }

        /// Returns the year number in the current era for this date
        #[diplomat::rust_link(icu::calendar::Date::year, FnInStruct)]
        pub fn year_in_era(&self) -> i32 {
            self.0.year().number
        }

        /// Returns the era for this date,
        #[diplomat::rust_link(icu::Date::year, FnInStruct)]
        #[diplomat::rust_link(icu::types::Era, Struct, compact)]
        pub fn era(
            &self,
            write: &mut diplomat_runtime::DiplomatWriteable,
        ) -> DiplomatResult<(), ICU4XError> {
            let era = self.0.year().era;
            let result = write.write_str(&era.0).map_err(Into::into).into();
            write.flush();
            result
        }

        /// Returns the number of months in the year represented by this date
        #[diplomat::rust_link(icu::calendar::Date::months_in_year, FnInStruct)]
        pub fn months_in_year(&self) -> u8 {
            self.0.months_in_year()
        }

        /// Returns the number of days in the month represented by this date
        #[diplomat::rust_link(icu::calendar::Date::days_in_month, FnInStruct)]
        pub fn days_in_month(&self) -> u8 {
            self.0.days_in_month()
        }

        /// Returns the number of days in the year represented by this date
        #[diplomat::rust_link(icu::calendar::Date::days_in_year, FnInStruct)]
        pub fn days_in_year(&self) -> u32 {
            self.0.days_in_year()
        }

        /// Returns the [`ICU4XCalendar`] object backing this date
        #[diplomat::rust_link(icu::calendar::Date::calendar, FnInStruct)]
        #[diplomat::rust_link(icu::calendar::Date::calendar_wrapper, FnInStruct, hidden)]
        pub fn calendar(&self) -> Box<ICU4XCalendar> {
            Box::new(ICU4XCalendar(self.0.calendar_wrapper().clone()))
        }
    }
}

File: ffi/mvn/src/main/kotlin/src/main/kotlin/org/unicode/icu4x/Date.kt
package org.unicode.icu4x;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface DateLib: Library {
    fun icu4x_Date_destroy_mv1(handle: Pointer)
    fun icu4x_Date_from_iso_in_calendar_mv1(isoYear: Int, isoMonth: FFIUint8, isoDay: FFIUint8, calendar: Pointer): ResultPointerInt
    fun icu4x_Date_from_codes_in_calendar_mv1(eraCode: Slice, year: Int, monthCode: Slice, day: FFIUint8, calendar: Pointer): ResultPointerInt
    fun icu4x_Date_from_rata_die_mv1(rd: Long, calendar: Pointer): ResultPointerInt
    fun icu4x_Date_from_string_mv1(v: Slice, calendar: Pointer): ResultPointerInt
    fun icu4x_Date_to_calendar_mv1(handle: Pointer, calendar: Pointer): Pointer
    fun icu4x_Date_to_iso_mv1(handle: Pointer): Pointer
    fun icu4x_Date_to_rata_die_mv1(handle: Pointer): Long
    fun icu4x_Date_day_of_year_mv1(handle: Pointer): FFIUint16
    fun icu4x_Date_day_of_month_mv1(handle: Pointer): FFIUint8
    fun icu4x_Date_day_of_week_mv1(handle: Pointer): Int
    fun icu4x_Date_weekday_mv1(handle: Pointer): Int
    fun icu4x_Date_ordinal_month_mv1(handle: Pointer): FFIUint8
    fun icu4x_Date_month_code_mv1(handle: Pointer, write: Pointer): Unit
    fun icu4x_Date_month_number_mv1(handle: Pointer): FFIUint8
    fun icu4x_Date_month_is_leap_mv1(handle: Pointer): Byte
    fun icu4x_Date_era_year_or_related_iso_mv1(handle: Pointer): Int
    fun icu4x_Date_extended_year_mv1(handle: Pointer): Int
    fun icu4x_Date_era_mv1(handle: Pointer, write: Pointer): Unit
    fun icu4x_Date_months_in_year_mv1(handle: Pointer): FFIUint8
    fun icu4x_Date_days_in_month_mv1(handle: Pointer): FFIUint8
    fun icu4x_Date_days_in_year_mv1(handle: Pointer): FFIUint16
    fun icu4x_Date_calendar_mv1(handle: Pointer): Pointer
}
/** An ICU4X Date object capable of containing a date for any calendar.
*
*See the [Rust documentation for `Date`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html) for more information.
*/
class Date internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
)  {

    internal class DateCleaner(val handle: Pointer, val lib: DateLib) : Runnable {
        override fun run() {
            lib.icu4x_Date_destroy_mv1(handle)
        }
    }

    companion object {
        internal val libClass: Class<DateLib> = DateLib::class.java
        internal val lib: DateLib = Native.load("icu4x", libClass)
        @JvmStatic
        
        /** Creates a new [Date] representing the ISO date
        *given but in a given calendar
        *
        *See the [Rust documentation for `new_from_iso`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.new_from_iso) for more information.
        */
        fun fromIsoInCalendar(isoYear: Int, isoMonth: UByte, isoDay: UByte, calendar: Calendar): Result<Date> {
            
            val returnVal = lib.icu4x_Date_from_iso_in_calendar_mv1(isoYear, FFIUint8(isoMonth), FFIUint8(isoDay), calendar.handle);
            if (returnVal.isOk == 1.toByte()) {
                val selfEdges: List<Any> = listOf()
                val handle = returnVal.union.ok 
                val returnOpaque = Date(handle, selfEdges)
                CLEANER.register(returnOpaque, Date.DateCleaner(handle, Date.lib));
                return returnOpaque.ok()
            } else {
                return CalendarErrorError(CalendarError.fromNative(returnVal.union.err)).err()
            }
        }
        @JvmStatic
        
        /** Creates a new [Date] from the given codes, which are interpreted in the given calendar system
        *
        *An empty era code will treat the year as an extended year
        *
        *See the [Rust documentation for `try_new_from_codes`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.try_new_from_codes) for more information.
        *
        *See the [Rust documentation for `try_from_str`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.Month.html#method.try_from_str) for more information.
        */
        fun fromCodesInCalendar(eraCode: String, year: Int, monthCode: String, day: UByte, calendar: Calendar): Result<Date> {
            val (eraCodeMem, eraCodeSlice) = PrimitiveArrayTools.borrowUtf8(eraCode)
            val (monthCodeMem, monthCodeSlice) = PrimitiveArrayTools.borrowUtf8(monthCode)
            
            val returnVal = lib.icu4x_Date_from_codes_in_calendar_mv1(eraCodeSlice, year, monthCodeSlice, FFIUint8(day), calendar.handle);
            if (returnVal.isOk == 1.toByte()) {
                val selfEdges: List<Any> = listOf()
                val handle = returnVal.union.ok 
                val returnOpaque = Date(handle, selfEdges)
                CLEANER.register(returnOpaque, Date.DateCleaner(handle, Date.lib));
                if (eraCodeMem != null) eraCodeMem.close()
                if (monthCodeMem != null) monthCodeMem.close()
                return returnOpaque.ok()
            } else {
                return CalendarErrorError(CalendarError.fromNative(returnVal.union.err)).err()
            }
        }
        @JvmStatic
        
        /** Creates a new [Date] from the given Rata Die
        *
        *See the [Rust documentation for `from_rata_die`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.from_rata_die) for more information.
        */
        fun fromRataDie(rd: Long, calendar: Calendar): Result<Date> {
            
            val returnVal = lib.icu4x_Date_from_rata_die_mv1(rd, calendar.handle);
            if (returnVal.isOk == 1.toByte()) {
                val selfEdges: List<Any> = listOf()
                val handle = returnVal.union.ok 
                val returnOpaque = Date(handle, selfEdges)
                CLEANER.register(returnOpaque, Date.DateCleaner(handle, Date.lib));
                return returnOpaque.ok()
            } else {
                return CalendarErrorError(CalendarError.fromNative(returnVal.union.err)).err()
            }
        }
        @JvmStatic
        
        /** Creates a new [Date] from an IXDTF string.
        *
        *See the [Rust documentation for `try_from_str`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.try_from_str) for more information.
        */
        fun fromString(v: String, calendar: Calendar): Result<Date> {
            val (vMem, vSlice) = PrimitiveArrayTools.borrowUtf8(v)
            
            val returnVal = lib.icu4x_Date_from_string_mv1(vSlice, calendar.handle);
            if (returnVal.isOk == 1.toByte()) {
                val selfEdges: List<Any> = listOf()
                val handle = returnVal.union.ok 
                val returnOpaque = Date(handle, selfEdges)
                CLEANER.register(returnOpaque, Date.DateCleaner(handle, Date.lib));
                if (vMem != null) vMem.close()
                return returnOpaque.ok()
            } else {
                return Rfc9557ParseErrorError(Rfc9557ParseError.fromNative(returnVal.union.err)).err()
            }
        }
    }
    
    /** Convert this date to one in a different calendar
    *
    *See the [Rust documentation for `to_calendar`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.to_calendar) for more information.
    */
    fun toCalendar(calendar: Calendar): Date {
        
        val returnVal = lib.icu4x_Date_to_calendar_mv1(handle, calendar.handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Date(handle, selfEdges)
        CLEANER.register(returnOpaque, Date.DateCleaner(handle, Date.lib));
        return returnOpaque
    }
    
    /** Converts this date to ISO
    *
    *See the [Rust documentation for `to_iso`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.to_iso) for more information.
    */
    fun toIso(): IsoDate {
        
        val returnVal = lib.icu4x_Date_to_iso_mv1(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = IsoDate(handle, selfEdges)
        CLEANER.register(returnOpaque, IsoDate.IsoDateCleaner(handle, IsoDate.lib));
        return returnOpaque
    }
    
    /** Returns this date's Rata Die
    *
    *See the [Rust documentation for `to_rata_die`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.to_rata_die) for more information.
    */
    fun toRataDie(): Long {
        
        val returnVal = lib.icu4x_Date_to_rata_die_mv1(handle);
        return (returnVal)
    }
    
    /** Returns the 1-indexed day in the year for this date
    *
    *See the [Rust documentation for `day_of_year`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.day_of_year) for more information.
    */
    fun dayOfYear(): UShort {
        
        val returnVal = lib.icu4x_Date_day_of_year_mv1(handle);
        return (returnVal.toUShort())
    }
    
    /** Returns the 1-indexed day in the month for this date
    *
    *See the [Rust documentation for `day_of_month`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.day_of_month) for more information.
    */
    fun dayOfMonth(): UByte {
        
        val returnVal = lib.icu4x_Date_day_of_month_mv1(handle);
        return (returnVal.toUByte())
    }
    
    /** Returns the day in the week for this day
    *
    *This is *not* the day of the week, an ordinal number that is locale
    *dependent.
    *
    *See the [Rust documentation for `day_of_week`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.day_of_week) for more information.
    */
    fun dayOfWeek(): Weekday {
        
        val returnVal = lib.icu4x_Date_day_of_week_mv1(handle);
        return (Weekday.fromNative(returnVal))
    }
    
    /** Returns the day in the week for this day
    *
    *See the [Rust documentation for `weekday`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.weekday) for more information.
    */
    fun weekday(): Weekday {
        
        val returnVal = lib.icu4x_Date_weekday_mv1(handle);
        return (Weekday.fromNative(returnVal))
    }
    
    /** Returns 1-indexed number of the month of this date in its year
    *
    *Note that for lunar calendars this may not lead to the same month
    *having the same ordinal month across years; use `month_code` if you care
    *about month identity.
    *
    *See the [Rust documentation for `month`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.month) for more information.
    *
    *See the [Rust documentation for `ordinal`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.MonthInfo.html#structfield.ordinal) for more information.
    */
    fun ordinalMonth(): UByte {
        
        val returnVal = lib.icu4x_Date_ordinal_month_mv1(handle);
        return (returnVal.toUByte())
    }
    
    /** Returns the month code for this date. Typically something
    *like "M01", "M02", but can be more complicated for lunar calendars.
    *
    *See the [Rust documentation for `code`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.Month.html#method.code) for more information.
    *
    *See the [Rust documentation for `standard_code`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.MonthInfo.html#structfield.standard_code) for more information.
    *
    *Additional information: [1](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.month)
    */
    fun monthCode(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.icu4x_Date_month_code_mv1(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** Returns the month number of this month.
    *
    *See the [Rust documentation for `number`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.Month.html#method.number) for more information.
    */
    fun monthNumber(): UByte {
        
        val returnVal = lib.icu4x_Date_month_number_mv1(handle);
        return (returnVal.toUByte())
    }
    
    /** Returns whether the month is a leap month.
    *
    *See the [Rust documentation for `is_leap`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.Month.html#method.is_leap) for more information.
    */
    fun monthIsLeap(): Boolean {
        
        val returnVal = lib.icu4x_Date_month_is_leap_mv1(handle);
        return (returnVal > 0)
    }
    
    /** Returns the year number in the current era for this date
    *
    *For calendars without an era, returns the related ISO year.
    *
    *See the [Rust documentation for `era_year_or_related_iso`](https://docs.rs/icu/2.1.1/icu/calendar/types/enum.YearInfo.html#method.era_year_or_related_iso) for more information.
    *
    *Additional information: [1](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.EraYear.html#structfield.year), [2](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.CyclicYear.html#structfield.related_iso), [3](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.year)
    */
    fun eraYearOrRelatedIso(): Int {
        
        val returnVal = lib.icu4x_Date_era_year_or_related_iso_mv1(handle);
        return (returnVal)
    }
    
    /** Returns the extended year, which can be used for
    *
    *This year number can be used when you need a simple numeric representation
    *of the year, and can be meaningfully compared with extended years from other
    *eras or used in arithmetic.
    *
    *See the [Rust documentation for `extended_year`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.extended_year) for more information.
    */
    fun extendedYear(): Int {
        
        val returnVal = lib.icu4x_Date_extended_year_mv1(handle);
        return (returnVal)
    }
    
    /** Returns the era for this date, or an empty string
    *
    *See the [Rust documentation for `era`](https://docs.rs/icu/2.1.1/icu/calendar/types/struct.EraYear.html#structfield.era) for more information.
    *
    *Additional information: [1](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.year)
    */
    fun era(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.icu4x_Date_era_mv1(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** Returns the number of months in the year represented by this date
    *
    *See the [Rust documentation for `months_in_year`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.months_in_year) for more information.
    */
    fun monthsInYear(): UByte {
        
        val returnVal = lib.icu4x_Date_months_in_year_mv1(handle);
        return (returnVal.toUByte())
    }
    
    /** Returns the number of days in the month represented by this date
    *
    *See the [Rust documentation for `days_in_month`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.days_in_month) for more information.
    */
    fun daysInMonth(): UByte {
        
        val returnVal = lib.icu4x_Date_days_in_month_mv1(handle);
        return (returnVal.toUByte())
    }
    
    /** Returns the number of days in the year represented by this date
    *
    *See the [Rust documentation for `days_in_year`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.days_in_year) for more information.
    */
    fun daysInYear(): UShort {
        
        val returnVal = lib.icu4x_Date_days_in_year_mv1(handle);
        return (returnVal.toUShort())
    }
    
    /** Returns the [Calendar] object backing this date
    *
    *See the [Rust documentation for `calendar`](https://docs.rs/icu/2.1.1/icu/calendar/struct.Date.html#method.calendar) for more information.
    */
    fun calendar(): Calendar {
        
        val returnVal = lib.icu4x_Date_calendar_mv1(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Calendar(handle, selfEdges)
        CLEANER.register(returnOpaque, Calendar.CalendarCleaner(handle, Calendar.lib));
        return returnOpaque
    }

}